### DISCUSSION TITLE
How to create relation in mongoose

### DISCUSSION THREAD

**Discussion by kansson** [doug-martin/nestjs-query#1179] [General]
Hello, I have read the docs for relations with mongoose but I'm lost. I have created the relation between two DTOs but how do I actually link them together in GraphQL? I can create two individual entries but not figure out how to link them together. What mutations should I use and in what order?

Consider these two entries, `Story` and `Comment`. One of each can be created with `createOne(Story/Comment)` but when I try to use `addCommentsToStory` it returns `AddRelations not supported for virtual relation comments`. What am I doing wrong? 

Thanks!

**Comment by smolinari** [comment]
What does your code look like for the two entities? 

Also, this should be in discussions, as it isn't clear it is an issue. Converting.

Scott

**Comment by kansson** [comment]
Sorry for posting this as an issue and not a discussion. These are my two entities and their DTOs. 
<details>
  <summary>story.entity.ts</summary>

  ```js
  import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
  import { Document } from 'mongoose';
  
  @Schema({ collection: 'stories' })
  export class StoryEntity extends Document {
    @Prop({ required: true })
    title!: string;
  
    @Prop({ required: true })
    text!: string;
  }
  
  export const StorySchema = SchemaFactory.createForClass(StoryEntity);
  StorySchema.virtual('comments', {
    ref: 'CommentEntity',
    localField: '_id',
    foreignField: 'story',
  });
  ```

</details>
<details>
  <summary>story.dto.ts</summary>

  ```js
  import {
    CursorConnection,
    FilterableField,
    IDField,
    KeySet,
  } from '@nestjs-query/query-graphql';
  import { ID, ObjectType } from '@nestjs/graphql';
  import { CommentDTO } from 'src/comments/comment.dto';
  
  @ObjectType('Story')
  @KeySet(['id'])
  @CursorConnection('comments', () => CommentDTO, { disableRemove: true })
  export class StoryDTO {
    @IDField(() => ID)
    id!: string;
  
    @FilterableField()
    title!: string;
  
    @FilterableField()
    text!: string;
  }
  ```

</details>
<details>
  <summary>comment.entity.ts</summary>

  ```js
  import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
  import { Document, SchemaTypes, Types } from 'mongoose';
  
  @Schema({ collection: 'comments' })
  export class CommentEntity extends Document {
    @Prop({ type: SchemaTypes.ObjectId, ref: 'StoryEntity', required: true })
    story!: Types.ObjectId;
  
    @Prop({ required: true })
    text!: string;
  }
  
  export const CommentSchema = SchemaFactory.createForClass(CommentEntity);
  ```
</details>
<details>
  <summary>comment.dto.ts</summary>

  ```js
  import {
    FilterableField,
    IDField,
    KeySet,
    Relation,
  } from '@nestjs-query/query-graphql';
  import { ID, ObjectType } from '@nestjs/graphql';
  import { StoryDTO } from 'src/stories/story.dto';
  
  @ObjectType('Comment')
  @KeySet(['id'])
  @Relation('story', () => StoryDTO, { disableRemove: true })
  export class CommentDTO {
    @IDField(() => ID)
    id!: string;
  
    @FilterableField()
    text!: string;
  }
  ```
</details>

My question is how I can create and link these two together through the relation.

**Reply by smolinari** [reply]
Where are you using the story model to "populate" the comments referencing the story?

Scott

**Reply by kansson** [reply]
I'm not sure if I'm doing that because I'm using the default `NestjsQueryGraphQLModule` to generate the CRUD routes and the `NestjsQueryMongooseModule` to handle the database interaction. Basically I followed the get started example and the one to many example for mongoose in the docs but with my own entities and DTOs.

I'm new to mongoose so is there something that I'm missing here?

Appreciate the help!

**Reply by kansson** [reply]
An update on what I found. If you create the story and comment separately with `createOne(Story/Comment)` and then set the story on the comment with `setStoryOnComment` the expected result is returned when querying the story with its comments. This works but is not ideal since the `story` field would be null on the comment until the set mutation has been called. Is there any way to do this in one mutation instead on the current solution? In some GraphQL CMSs you can `connect` a relation when creating the entity. Maybe this can be implemented somehow by modifying the DTOs?

Thanks

**Reply by smolinari** [reply]
Yes, so, usually you'd have the story written and published and when someone is commenting on it, the storyId is known. When the comment is written and should be saved, the storyId is also saved at that same time. Right? In other words, there would have to be a story first, before there can be any commenting. You don't need to have both created at the same time or at least you shouldn't. 

Scott

**Reply by kansson** [reply]
Found out that you can have a custom `CreateDTOClass` to include the story when creating the comment as peer the mongoose example. https://github.com/doug-martin/nestjs-query/tree/master/examples/mongoose
